File: frontend/src/api/dataSources.ts
/**
 * 数据源接口（M6 / T6.1–T6.3）。
 *
 * 一个数据源 = "定期去看某处有没有新内容"。当前支持 RSS 订阅与单页 HTML。
 */

import { request } from './client'
import type { components } from './schema'

/** 界面支持的形式。后端将来加种类时，见下面 `kind` 的处理。 */
export type SourceKind = 'rss' | 'html'

type DataSourceOut = Required<components['schemas']['DataSourceOut']>

/**
 * 数据源：契约来自后端的 OpenAPI。
 *
 * `kind` **显式放宽成 `string`**：后端加一种数据源（比如 WebDAV）时，
 * 老前端不该因为"读到一个没见过的 kind"而崩——它只需要显示一个名字。
 * 收窄成 `SourceKind` 会让新种类在上线那天变成类型错误，而那时前端还没发版。
 */
export type DataSource = Omit<DataSourceOut, 'kind'> & { kind: SourceKind | string }

export interface SyncResult {
  task_id: string | null
  source_id: string
  fetched: number
  created: number
  duplicates: number
  /** 服务端回了 304：源没有任何变化。 */
  not_modified: boolean
  errors: string[]
}

export function listDataSources(kbId: string): Promise<{ items: DataSource[] }> {
  return request(`/knowledge-bases/${kbId}/data-sources`)
}

export function createDataSource(
  kbId: string,
  payload: { kind: SourceKind; name: string; url: string; max_items?: number | null },
): Promise<DataSource> {
  return request(`/knowledge-bases/${kbId}/data-sources`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function setDataSourceEnabled(id: string, enabled: boolean): Promise<DataSource> {
  return request(`/data-sources/${id}/enabled?enabled=${enabled}`, { method: 'PATCH' })
}

export function deleteDataSource(id: string): Promise<void> {
  return request(`/data-sources/${id}`, { method: 'DELETE' })
}

/**
 * 拉取一次。
 *
 * `wait=true` 会同步跑完并返回统计（界面用这个——用户点了"立即拉取"
 * 就该立刻看到"抓到了几条"，而不是一个任务号）。
 * 后台定时任务走入队那条路。
 */
export function syncDataSource(id: string, wait = true): Promise<SyncResult> {
  return request(`/data-sources/${id}/sync?wait=${wait}`, { method: 'POST' })
}

File: backend/app/api/v1/data_sources.py
"""数据源端点（M6 / T6.1–T6.3）。

四个动作：**登记、列出、启停、拉取**。

**为什么拉取要分同步/异步两条**：
- ``POST /data-sources/{id}/sync`` 入队（默认）——一个源几十条，每条都要解析
  与向量化，同步做会把 HTTP 请求挂几分钟；
- ``POST ...?wait=true`` 立刻做完并返回结果——给"我就想立刻看到抓到了什么"用，
  测试也走这条。

**权限**：读写。拉取会往知识库里加文档，只读密钥不该能做。
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from app.api.auth import WRITE, check_kb_scope, require_read, require_write
from app.api.v1.schemas import (
    DataSourceCreateIn,
    DataSourceListOut,
    DataSourceOut,
    SyncResultOut,
)
from app.core.services import Services, get_services
from app.services.api_key import Caller

router = APIRouter(tags=["data-sources"])


def _out(services: Services, record) -> DataSourceOut:  # type: ignore[no-untyped-def]
    return DataSourceOut(
        id=record.id,
        knowledge_base_id=record.knowledge_base_id,
        kind=record.kind.value,
        name=record.name,
        url=str(record.config.get("url") or ""),
        max_items=int(record.config.get("max_items") or 0) or None,
        enabled=record.enabled,
        etag=record.etag,
        last_pulled_at=record.last_pulled_at,
    )


@router.get(
    "/knowledge-bases/{kb_id}/data-sources",
    response_model=DataSourceListOut,
    summary="某知识库的数据源",
)
def list_data_sources(
    kb_id: str,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_read)],
) -> DataSourceListOut:
    check_kb_scope(services, caller, [kb_id])
    return DataSourceListOut(
        items=[_out(services, item) for item in services.sources.list(kb_id)]
    )


@router.post(
    "/knowledge-bases/{kb_id}/data-sources",
    response_model=DataSourceOut,
    status_code=status.HTTP_201_CREATED,
    summary="登记数据源（HTML / RSS）",
)
def create_data_source(
    kb_id: str,
    payload: DataSourceCreateIn,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_write)],
) -> DataSourceOut:
    """登记一个数据源。

    **登记不等于拉取**：刚登记完不会立刻有文档，要么等定时任务，
    要么显式点一次"拉取"。界面上要说清这一点，否则用户会以为登记完就有了。
    """
    check_kb_scope(services, caller, [kb_id], need=WRITE)
    from app.models.enums import DataSourceKind

    record = services.sources.create(
        knowledge_base_id=kb_id,
        kind=DataSourceKind(payload.kind),
        name=payload.name,
        url=payload.url,
        max_items=payload.max_items,
    )
    return _out(services, record)


@router.patch(
    "/data-sources/{source_id}/enabled",
    response_model=DataSourceOut,
    summary="启用 / 停用数据源",
)
def toggle_data_source(
    source_id: str,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_write)],
    enabled: bool = Query(description="是否启用"),
) -> DataSourceOut:
    record = services.sources.get(source_id)
    check_kb_scope(services, caller, [record.knowledge_base_id], need=WRITE)
    return _out(services, services.sources.set_enabled(source_id, enabled=enabled))


@router.delete(
    "/data-sources/{source_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="删除数据源（已抓取的文档保留）",
)
def delete_data_source(
    source_id: str,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_write)],
) -> None:
    """删数据源，**已抓取的文档保留**——它们是知识库的正式内容，可能已被引用。
    停掉订阅不等于要撤销已经收集的资料。
    """
    record = services.sources.get(source_id)
    check_kb_scope(services, caller, [record.knowledge_base_id], need=WRITE)
    services.sources.delete(source_id)


@router.post(
    "/data-sources/{source_id}/sync",
    response_model=SyncResultOut,
    summary="拉取一次（默认入队；wait=true 立刻做完）",
)
def sync_data_source(
    source_id: str,
    services: Annotated[Services, Depends(get_services)],
    caller: Annotated[Caller, Depends(require_write)],
    wait: bool = Query(default=False, description="true 则同步做完并返回统计"),
) -> SyncResultOut:
    """拉一次。

    ``wait=false``（默认）只入队并返回任务 id；``wait=true`` 直接跑完，
    返回"取回几条、新入库几条、重复几条、失败几条"。
    """
    record = services.sources.get(source_id)
    check_kb_scope(services, caller, [record.knowledge_base_id], need=WRITE)
    if not wait:
        task_id = services.sources.sync_async(source_id)
        return SyncResultOut(task_id=task_id)

    outcome = services.sources.sync_now(source_id)
    return SyncResultOut(**outcome.as_dict())  # type: ignore[arg-type]
